provider/oci/provider_test.go

```diff
@@ -0,0 +1,211 @@
+	return cloud.NewCredential(cloud.HTTPSigAuthType, map[string]string{
+		"key":         ocitesting.PrivateKeyEncrypted,
+		"fingerprint": ocitesting.PrivateKeyEncryptedFingerprint,
+		"pass-phrase": ocitesting.PrivateKeyPassphrase,
+		"tenancy":     "fake",
+		"user":        "fake",
+		"region":      "us-ashburn-1",
+	})
+}
+
+func (s *credentialsSuite) SetUpTest(c *gc.C) {
+	s.FakeHomeSuite.SetUpTest(c)
+
+	s.provider = &oci.EnvironProvider{}
+	s.spec = fakeCloudSpec()
+}
+
+func (s *credentialsSuite) writeOCIConfig(c *gc.C, sections map[string]map[string]string) {
+	home := utils.Home()
+	sectionList := []string{}
+	for k, v := range sections {
+		pem_name := fmt.Sprintf(".oci/oci_api_key_%s.pem", k)
+		pem := filepath.Join(home, pem_name)
+		s.Home.AddFiles(c, testing.TestFile{
+			Name: pem_name,
+			Data: v["key"],
+		})
+		cfg := fmt.Sprintf(
+			singleSectionTemplate, k, v["fingerprint"],
+			pem, v["region"], v["pass-phrase"])
+		sectionList = append(sectionList, cfg)
+	}
+	s.Home.AddFiles(c, testing.TestFile{
+		Name: ".oci/config",
+		Data: strings.Join(sectionList, "\n")})
+}
+
+func (s *credentialsSuite) TestCredentialSchemas(c *gc.C) {
+	envtesting.AssertProviderAuthTypes(c, s.provider, "httpsig")
+}
+
+func (s *credentialsSuite) TestUserPassCredentialsValid(c *gc.C) {
+	envtesting.AssertProviderCredentialsValid(c, s.provider, "httpsig", map[string]string{
+		"user":        "fake",
+		"tenancy":     "fake",
+		"key":         ocitesting.PrivateKeyEncrypted,
+		"pass-phrase": ocitesting.PrivateKeyPassphrase,
+		"fingerprint": ocitesting.PrivateKeyEncryptedFingerprint,
+		"region":      "us-phoenix-1",
+	})
+}
+
+func (s *credentialsSuite) TestPassphraseHiddenAttributes(c *gc.C) {
+	envtesting.AssertProviderCredentialsAttributesHidden(c, s.provider, "httpsig", "pass-phrase")
+}
+
+func (s *credentialsSuite) TestDetectCredentialsNotFound(c *gc.C) {
+	_, err := s.provider.DetectCredentials()
+	c.Assert(errors.Cause(err), jc.Satisfies, os.IsNotExist)
+}
+
+func (s *credentialsSuite) TestDetectCredentials(c *gc.C) {
+	cfg := map[string]map[string]string{
+		"DEFAULT": map[string]string{
+			"fingerprint": ocitesting.PrivateKeyEncryptedFingerprint,
+			"pass-phrase": ocitesting.PrivateKeyPassphrase,
+			"key":         ocitesting.PrivateKeyEncrypted,
+			"region":      "us-phoenix-1",
+		},
+	}
+	s.writeOCIConfig(c, cfg)
+	creds, err := s.provider.DetectCredentials()
+	c.Assert(err, gc.IsNil)
+	c.Assert(len(creds.AuthCredentials), gc.Equals, 1)
+	c.Assert(creds.DefaultRegion, gc.Equals, "us-phoenix-1")
+}
+
+func (s *credentialsSuite) TestDetectCredentialsWrongPassphrase(c *gc.C) {
+	cfg := map[string]map[string]string{
+		"DEFAULT": map[string]string{
+			"fingerprint": ocitesting.PrivateKeyEncryptedFingerprint,
+			"pass-phrase": "bogus",
+			"key":         ocitesting.PrivateKeyEncrypted,
+			"region":      "us-phoenix-1",
+		},
+	}
+	s.writeOCIConfig(c, cfg)
+	_, err := s.provider.DetectCredentials()
+	c.Assert(err, gc.NotNil)
```

Comment:
Do we want to use gc.ErrorMatches, and assert that the error string we're getting is something that a user has a chance to understand?

Reply:
Sure.